lib/pbench/server/database/models/datasets.py

```diff
@@ -671,6 +671,28 @@ class Metadata(Database.Base):
 
     __tablename__ = "dataset_metadata"
 
+    # +++ Standard Metadata keys:
+    # Lowercase keys are for client use, while uppercase keys are reserved for
+    # internal use.
```

Comment:
OK, so the comment could be improved. As for the naming, I just started with all-caps for the original internal keys. I decided I wanted the client-visible keys to be distinct, and figured lowercasing them was a natural distinction.

The part I haven't quite figured out is how I want to map end user metadata on top of server metadata (like DELETED, INDEXED), and dashboard "standard" metadata (like SEEN, SAVED). This either means dropping the idea of an ENUM-like reserved keyword space (which is why I didn't use an ENUM in the first place), or mapping them all through something like a `USER` JSON object which could contain any keys the client (e.g. dashboard) wanted to allow (which has the advantage of enforcing a separate namespace at that second level).

In the latter model, I really ought to just make a `Metadata.KEYS` ENUM, which could have a "private" property that disallows API `PUT` operations, and that would simplify a lot of this code and avoid the confusion.

Reply:
I like the two-level approach, with the top level being "server" vs. "client".  We probably would be well-advised not to do too much with the "client" sub-tree:  from the perspective that the Dashboard won't be our only client, if the server "knows" too much about the client sub-tree, then that becomes part of the server's interface, which I suspect that we don't want.  (Or, maybe we just need a third branch at the root of the tree....)

Reply:
I went back and forth on this one.

On one hand, I could define and control server-owned metadata with a key validation list (mostly for protection against external updates), but allow the clients to define any other keys they want.

I chose to promote dashboard metadata to a special class, and collect anything else under an arbitrary JSON valued "user" keyword. So the metadata key itself is reserved, but the server will just store and retrieve the value without any special interpretation. This keeps the "arbitrary" user metadata out of the first-level key namespace in case we want to add more later, at the expense of moving user-owned metadata down a level in the tree.

Reply:
I just had the thought that a way to treat the client-controlled keys as, perhaps, a cleaner second-level namespace would be instead of having `"user"` be a single `JSON` document, to allow clients to specify `user.xxx` keys where the server implicitly manages the name following the period as a key within the `user` JSON. So if we saw the name `user.x` we'd set or return the value of the key "x" in the "user" Metadata value.

Reply:
I'm not sure I follow.  I mean, I like the idea of a partitioned space, which one can implement by using prefixed names, but I think what you're suggesting will require the server to parse the names, which seems...unsatisfying.  That is, since the input is JSON, what's wrong with just having nested objects -- they shouldn't be hard for the Javascript client to produce, and they are easy for the Python server (once the JSON package, or Flask, or whatever renders the JSON into a native object has done its work).

Reply:
`"user.contact": "junk"` provides a level of isolation that protects the internal implementation as well as helping to protect the shared namespace: we don't have to worry about what happens if someone says `"user": "junk"`, because that's not a defined key. Is that an error? Is it legitimate to replace the entire shared "user" sub-document?

Reply:
All top-level keys (i.e., the ones in the `metadata` `dict`) are reserved to the server; if we get a request to set user metadata, it comes in key-value pairs, so the result is
```
  {
    "metadata": {
      "user": {
        "<user-key1>": "<value1>"
      }
    }
  }
```
The server could have keys which are peers to `"metadata"` or which are peers to `"user"` (e.g., `"server"`) or which are under `"server"` as we see fit; but the client (which makes requests via the API) cannot access anything above `"user"` (except if we provide a `"client"` which is a peer to `"user"`).  So, there are no conflicts with server keys.

If a duly-authorized API request wants to change a value for a key, then we can locate that key in the client- or user-namespace and set the new value; if they replace a subtree at that point, that's on them.

If we want to have two namespaces (e.g., just `"client"` and `"server"`) instead of three (removing `"user"`), then it's on the client to maintain their own namespace (e.g., by requesting a `"user"` key which we would place under "`client`"). If some other client request sets a new value for that key...well...that's not really our problem.

However, it _would_ behoove us to make it easy for multiple clients to get along.  Instead of a generic `"client"` key for the client's namespace, we could entertain the notion of having the client provide its own top-level key...but, then we're back to trying to defend against conflicts with server keys.  We could use a prefix, or add another level, to keep the client keys corralled...so the client would give us their client-key and the metadata key, and we would add a prefix to the client-key or look for it in the client-namespace and then look up the metadata key there.

Reply:
This is starting to feel over-engineered, like we need a bunch of documentation to explain the hierarchy which may never actually be used. But I'll give this a try ...

OK, so, first level namespace is the reserved server metadata, which is read-only for clients, with a `client` subdocument that's theirs to play with, even where we may initially populate default values for convenience (which is currently just setting `seen` and `saved` to False on dataset upload).

Any attempt to PUT a value must be in the `client` subdocument. If they 

```
PUT /dataset/metadata
{
  ...,
  "metadata": {
    "client": blah
  }
}
```

Then they overwrite it all, and it's on them. (We'll audit log that eventually, so an admin can see who did it; and of course it needs to be someone with UPDATE access to that dataset in the first place.)

But if we currently have

```
"metadata": {
  "deletion": "2222-13-42",
  "archived": True,
  "client": {
    "seen": True,
    "saved": False,
    "contact": "me@example.com"
  }
}
```

and they do

```
PUT /datasets/metadata
{
  "metadata": {
    "client": {
      "mykey": "my precious",
      "contact": "you@example.com"
  }
}
```

we'll *only* update `mykey` and `contact` in the second level, resulting in

```
"metadata": {
  "deletion": "2222-13-42",
  "archived": True,
  "client": {
    "seen": True,
    "saved": False,
    "contact": "you@example.com",
    "mykey": "my precious"
  }
}
```

This probably means that I should parse the GET paths to allow `&metadata=delete&metadata=client.contact`, and I guess I'm doing keyword validation only on the first level, which means "saved" and "seen" lose their keywordness. But I do kinda like the idea of making only second-level keys user-updateable.

Reply:
This is a brain dump; read at your own risk, but thoughts and discussion are desired... and I think I'm about to start coding ... something ... soonish ... as it gels a bit more in my mind.

I'm reconsidering that. I think that for `PUT` it makes sense to support `{"client": {"contact": "me", "saved": True}}` but I'm not convinced that the explicit two-level path makes sense on a "get": e.g., `"metadata": {"client": ["contact", "saved"]}` in the `datasets/detail` and `datasets/list`; for example, to mix top-level and second-level keys, we'd need something more like `"metadata": ["deleted", {"client": ["contact", "saved"]}]` which isn't good JSON (strings and objects mixed in a list), and is messier to construct and read. More importantly I don't want to require or parse `&metadata={"client":["contact","saved"]}` for a `GET /datasets/metadata` nor do I really want to give up on `GET` and put the specification in a request JSON payload.

I think instead that a flattened namespace like `client.saved` makes a lot more sense for the "get" cases (both query parameter and `"metadata"` JSON specification), which suggests that we should be consistent for the `PUT` to update that data.

In other words,

```
POST /datasets/list
{
  "user": "me",
  "controller": "host",
  "name": "fred",
  "metadata": ["deleted", "client.saved", "client.seen"]
}
```

```
PUT /datasets/metadata
{
  "controller": "host",
  "name": "fred",
  "metadata": {
    "client.saved": False,
    "client.seen": True
    "client.contact": "me@example.com"
  }
}
```

and

`GET /datasets/metadata?controller=host&name=fred&metadata=deleted&metadata=client.seen`

I don't know whether it's good or bad that this brings us back to the possibility of straightforward keyword validation on the second level, because I do think we want to allow clients to create arbitrary second level client/user keys; so probably I only validate the first segment... *although* there's something to be said for allowing clients to define their own first-level key, so we effectively define "metadata key" as a dot-separated name where the first segment doesn't overlap the reserved server first-level metadata keys.

I could minimize that reserved space by demoting ours to `server.deleted`, `server.archived` and so forth, which might clarify the namespace a bit even though it means more change to existing server code... right now I'm inclined to restrict the namespace layering to the API and avoid that overhead internally, which means retaining the internal server keys as top-level `Metadata` keys in the DB.

This does however mean that once we stabilize and release, we'll need to handle this anyway in order to add reserved keys without potentially breaking a client. One "cheat" might be to treat "server" as a reserved namespace, and *name* the server keys as `server.deleted` and such, but still store them as first-level `Metadata` rows rather than packing/unpacking in a JSON `server` document. Then we can arbitrarily add new server namespace keys without conflicting with user keys and avoid the packing outside the API. (Now I need to sit on tat a bit and decide whether it makes me feel bad. 🙄 )

Then again, we could simply do this for all metadata keys, enforcing the dot-prefix namespace, so `client.seen` and `client.saved` would both be independent metadata table rows instead of combined in a `client` metadata key JSON document. That has advantages and disadvantages... I rather like the simplicity of `&metadata=client` returning the whole document, which falls out pretty naturally from the nested document model but means a more complicated query and packing operation if we're actually fetching multiple metadata rows and stitching them together.

Sadly, either way will mean breaking legacy test gold files again by changing the rows in the metadata table.

I'm tempted to try to get this PR done and worry about it later; however as this affects how the dashboard gets the metadata for the overview page, which was the whole point, it's probably best to resolve it now. This note is the result of my initial struggle over implementation details, although I haven't actually coded anything yet. I'm going to try roughing this out as a separate commit, and we'll see what people think.

Reply:
I think that we have perhaps three aspects to this problem, and I suspect that if we avoid conflating them we'll come to a satisfactory conclusion faster.
 - I think we have the question of how we represent the various pieces of data and metadata in the database.  Note that different parts of the namespace (in fact, individual items) might have different representations in the database.  For instance, some server metadata items might be in the Dataset; other server metadata might be in rows in an auxiliary table; client metadata might be a JSON blob in a single row/column/cell-thingie; etc.
 - I think we have the question of how a client request refers to a particular metadata item.  Here we are free to impose context, to consider references to be relative to some subordinate node in the conceptual tree, etc.  We are free to use JSON objects or flat, segmented names, et al.  And, we have the _option_ (although I think you've already discarded it...) of using different idioms for `GET` and `PUT`.
 - And, finally, we have the conceptual tree, which is where I would start my thought processes.

Conceptually, I would propose that we have a tree or subtree rooted at a node called `metadata`.  Under that node, I would propose a `server` node and at least one "client" node.  For the moment, we can label that node `client`, although I think we should remain open to naming it anything which doesn't conflict with `server` or any other top-level key that we decide to create.  Under `client` we have the key-value pairs that we are trying to maintain on behalf of the client, such as `seen`, `saved`, and any others that the client sees fit to create (including keys which refer to objects or whole subtrees).  Conceptually, under `server`, we have the server-side metadata, such as `deleted`, `owner`, `access`, et al.  So far, so good (I hope...).

Now, when the client wants to `GET` metadata, they need to provide their client key (unless the target of the request is server-side metadata).  So long as we have only one client, this could be `client`, but we'll presumably want to be able to keep one client's metadata separate from another client's, should we ever want to support more than one.  Presumably, the client key would be part of the URI, although that's not a given, but, if so, it is something that we could add later.  The rest of the request comprises a list of keys.  These could be specified as a JSON string passed as a query parameter or as individual query parameters.  The advantage of the JSON string is that it can provide structural context (and it's easy enough to parse it into native types); the disadvantage is that it's ugly-looking.  Individual parameters, on the other hand, need to provide explicit pathing information (e.g., `server.deleted`, `client.seen`, `client.profile.user.formal_name.last`, etc.), because it will be rendered as a flat list, unlike the nested-object structure which JSON would afford.  But, one way or another, the server would receive a list of requested fields, and it would parse each of them into a tree traversal which could be used to locate the corresponding value in the conceptual tree.  In terms of where the traversal starts, we have options -- the simplest approach is to start under `metadata`, but we could consider other strategies, such as heuristic searches (e.g., looking under `server` first, then under the client key, then perhaps deeper).  Once we determine where in the conceptual tree the key points, then we can map that to the appropriate database access to retrieve the value.  Note of course that we don't necessarily have to traverse an actual tree: the tree is just a concept which guides how to map requested locations into suitable database queries.  The additional wrinkle is how to present the values back to the client.  Again, we can either instantiate the conceptual tree as a JSON object with the values filled in (perhaps using the input JSON as a template), or we can return a single-level object with segmented path-names as keys.

When the client wants to `PUT` metadata, we face the same challenge:  we need to get the list of key-value pairs from the request and map the keys to database accesses via the conceptual tree.  In this case, we have the advantage that the request carries a payload which could be nested-object JSON.  But, it could also be just a flat list of path references.  Either way, the result follows what we do for `GET`.

FWIW, both JSON and Python are fine with heterogenous arrays/lists, so we _could_ accept `"metadata": ["deleted", {"client": ["contact", "saved"]}]`.  Or, we could consider adding a keyword...something like `"metadata": { "keys":["deleted"], "client": { "keys": ["contact", "saved"]}}`, but I don't like that very much (and, we would have to come up with some way to make sure that `"keys"` didn't conflict with anything).  Or, we could do a fill-in-the-blanks object: `"metadata": {"deleted": None, {"client": {"contact": None, "saved": None}}}` where we replace the `None`'s (actually, any value which is not an object) with the metadata value corresponding to its key.  (This might be be kind of cool, actually, as it makes the `GET` and `PUT` forms directly analogous, even if the `GET` one has to be rendered as a string for the query parameter.)

Reply:
So, what you have now is really good.  There are some lingering issues with it, but I would prefer to discuss them elsewhere and close out this conversation.

However, before I do that, I have to point out that there's still a work item here:  now that you've revamped the keys, the code comment at lines 696-7 are stale.  😀